src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -1254,6 +1255,41 @@ theorem saddOverflow_eq {w : Nat} (x y : BitVec w) :
     simp
     omega
 
+theorem umulOverflow_eq {w : Nat} (x y : BitVec w) :
+    umulOverflow x y =
+      (0 < w && BitVec.twoPow (w * 2) w ≤ x.zeroExtend (w * 2) * y.zeroExtend (w * 2)) := by
+  simp [umulOverflow, ge_iff_le, truncate_eq_setWidth, le_def, toNat_twoPow, toNat_mul,
+    toNat_setWidth, Nat.mul_mod_mod, Nat.mod_mul_mod]
+  rcases w with _|w
+  · simp [of_length_zero, toInt_zero]
+  · simp only [show 0 < w + 1 by omega, decide_true, Bool.true_and, decide_eq_decide]
+    rw [Nat.mod_eq_of_lt BitVec.toNat_mul_toNat_lt, Nat.mod_eq_of_lt]
+    have := Nat.pow_lt_pow_of_lt (a := 2) (n := (w + 1)) (m := (w + 1) * 2)
+    omega
+
+theorem smulOverflow_eq {w : Nat} (x y : BitVec w) :
+    smulOverflow x y
+      = ((w ≠ 0) &&
+    ((signExtend (w * 2) (intMax w)).slt (signExtend (w * 2) x * signExtend (w * 2) y) ||
+      (signExtend (w * 2) x * signExtend (w * 2) y).slt (signExtend (w * 2) (intMin w)))) := by
+  simp only [smulOverflow]
+  rcases w with _|w
+  · simp [of_length_zero, toInt_zero]
+  · simp only [Nat.add_one_sub_one, ge_iff_le, ne_eq, show ¬(w + 1) = 0 by omega,
+    not_false_eq_true, decide_true, BitVec.slt, intMax, ofNat_eq_ofNat, toInt_mul, intMin,
+    Bool.true_and]
```

Comment:
Are all these needed here or can we drop some?

Reply:
Maybe at the cost of a simp in the end? At least, then we do not need to list them separately, as it is in the end.

Reply:
I tried to move as many as possible toward the end, but it seems that these are in fact necessary.